src/NServiceBus.Core.Analyzer/ForwardCancellationTokenAnalyzer.cs

```diff
@@ -0,0 +1,308 @@
+            ForwardCancellationTokenDiagnostic
+        );
+
+        class AnalysisTarget
+        {
+            public string TypeName { get; }
+            public string MethodName { get; }
+            public string ContextTypeName { get; }
+
+            public AnalysisTarget(string typeName, string methodName, string contextTypeName) =>
+                (TypeName, MethodName, ContextTypeName) = (typeName, methodName, contextTypeName);
+        }
+
+        public override void Initialize(AnalysisContext context) =>
+            context.WithDefaultSettings().RegisterCompilationStartAction(AnalyzeCompilationStart);
+
+        static void AnalyzeCompilationStart(CompilationStartAnalysisContext startContext)
+        {
+            var knownTypes = new KnownTypes(startContext.Compilation);
+            if (knownTypes.IsValid)
+            {
+                startContext.RegisterSyntaxNodeAction(context => AnalyzeSyntax(context, knownTypes), SyntaxKind.InvocationExpression);
+            }
+        }
+
+        class KnownTypes
+        {
+            public INamedTypeSymbol CancellationTokenType { get; }
+            public INamedTypeSymbol CancellableContextType { get; }
+            public ImmutableArray<INamedTypeSymbol> AnalysisTypes { get; }
+
+            public bool IsValid => CancellationTokenType != null && CancellableContextType != null;
+
+            public KnownTypes(Compilation compilation)
+            {
+                CancellationTokenType = compilation.GetTypeByMetadataName("System.Threading.CancellationToken");
+
+                CancellableContextType = compilation.GetTypeByMetadataName("NServiceBus.ICancellableContext");
+
+                AnalysisTypes = targets
+                    .Select(target =>
+                    {
+                        var type = compilation.GetTypeByMetadataName(target.TypeName);
+
+                        // Translate IBehavior<TFromContext, TToContext> to IBehavior<,>
+                        return type.IsGenericType ? type.ConstructUnboundGenericType() : type;
+                    })
+                    .ToImmutableArray();
+            }
+        }
+
+        static void AnalyzeSyntax(SyntaxNodeAnalysisContext context, KnownTypes knownTypes)
+        {
+            if (!(context.Node is InvocationExpressionSyntax invocation))
+            {
+                return;
+            }
+
+            if (!(invocation.Ancestors().OfType<MethodDeclarationSyntax>().FirstOrDefault() is MethodDeclarationSyntax callingMethodSyntax))
+            {
+                return;
+            }
+
+            if (!(context.SemanticModel.GetDeclaredSymbol(callingMethodSyntax) is IMethodSymbol callingMethod))
+            {
+                return;
+            }
+
+            if (!(callingMethod.Parameters.FirstOrDefault(param => IsCancellableContext(param.Type, knownTypes)) is IParameterSymbol contextParam))
+            {
+                return;
+            }
```

Comment:
@DavidBoike I had to go ahead and change this to using the semantic model, because removing the "does the containing type inherit from X" condition made `NoDiagnosticWhenNotNServiceBus` test fail, because the `IMadeUpContext` type was satisfying the conditions for it to be considered a cancellable context.

Reply:
Well fine. Starting from the method view, if that works, will do a lot less analysis on parameters in the first place.